### PR TITLE
Towards unifying fuzz targets and module level tests

### LINKED ISSUE #2377: Add shim between proptest and fuzzing
In `dkg2`, we've created a nice pattern where we define a test plan for various scenarios, use that to define unit tests, and then implement `arbitrary::Arbitrary` to allow the fuzzer to try different cases.

In other places, we use `proptest`, as a convenient kind of fuzzer with nice shrinking, for in crate tests.

It would be nice to avoid having to define strategies twice, for `proptest` and for `arbitrary`. We could instead define a "shim" which allows a test plan that implements `proptest::Arbitrary`, and has a way to be executed, to be exercised by a fuzzer.

This shouldn't be too hard, because the `proptest` crate has the ability to have randomness provided to it from some external source, so we could use the fuzzer to guide the proptest runner for that structure.

### PR DESCRIPTION
**Pull request by cronokirby**
Closes #2377.

I implement this idea in the math crate, with the idea of expanding it to other places, e.g. cryptography, in subsequent work.

The pain felt was that there could be a lot of duplication between fuzz testing code and property testing code, and that we were not doing particularly intelligent fuzzing, nor using property testing enough.

To alleviate both of these at once, I wanted a way to easily define property tests, and then have those drive fuzz testing as well.

The implementation here has each crate create a top level "fuzzing plan" which an then be run both as a fuzz target, and as a property test.

To make it easy to run fuzzing for that plan inside of a module level test, I cooked up a little fuzzing library, which we can easily tweak as needed. For now, it works well enough. It can generate about 400k samples a second, which seems more than ample to not be bottlenecked by the actual time to run whatever function you're testing, so I'm not worried about that.

<sup><a href="https://cursor.com/dashboard?tab=bugbot">Cursor Bugbot</a> reviewed your changes and found no issues for commit <u>2d89bb9</u></sup>

### CODE REVIEW

File: math/fuzz/fuzz_targets/main.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#![no_main]
+
+libfuzzer_sys::fuzz_target!(|input: &[u8]| {
+    use arbitrary::{Arbitrary, Unstructured};
```
Comment (patrick-ogrady) [inline math/fuzz/fuzz_targets/main.rs]: import outside of loop?

File: math/src/lib.rs [hunk 2]
```diff
@@ -19,5 +19,35 @@ commonware_macros::stability_scope!(BETA {
     pub mod poly;
 });
 
-#[cfg(test)]
-pub(crate) mod test;
+commonware_macros::stability_scope!(BETA {
+    #[cfg(any(test, feature = "fuzz"))]
+    pub(crate) mod test;
```
Comment (patrick-ogrady) [inline math/src/lib.rs:24]: Thus far, I've taken the approach that all testing infra should be `ALPHA`.

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# commonware-test
+
+[![Crates.io](https://img.shields.io/crates/v/commonware-test.svg)](https://crates.io/crates/commonware-test)
+[![Docs.rs](https://docs.rs/commonware-test/badge.svg)](https://docs.rs/commonware-test)
+
+Simplify testing and fuzzing of distributed systems.
+
+## Status
+
+`commonware-test` is **ALPHA** software and is not yet recommended for production use. Developers should expect breaking changes and occasional instability.
```
Comment (patrick-ogrady) [inline test/README.md]: Use the standard stability status line.

File: test/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+//! Simplify testing and fuzzing of distributed systems.
```
Comment (patrick-ogrady) [inline test/src/lib.rs]: IIRC we also add the `Status` line here?

File: invariants/Cargo.toml
Comment (patrick-ogrady) [inline invariants/Cargo.toml:1]: Wondering if we should just place `minifuzz` in `utils`? I suppose the idea is that we'll have more testing utilities that don't make sense to put anywhere else?

File: Cargo.toml [hunk 5]
```diff
@@ -18,6 +18,7 @@ members = [
     "runtime",
     "storage",
     "stream",
+    "test",
```
Comment (patrick-ogrady) [inline Cargo.toml]: I think we could call this "invariant"

Comment (patrick-ogrady) [reply Cargo.toml]: On the theme of performance

File: README.md [hunk 6]
```diff
@@ -50,6 +50,7 @@ _Sometimes, we opt to maintain software that is neither a primitive nor an examp
 * [docker](./docker): Dockerfiles used for cross-compilation and CI.
 * [macros](./macros/README.md): Augment the development of primitives with procedural macros.
 * [mcp](./mcp/README.md): Interact with the Commonware Library via MCP at https://mcp.commonware.xyz.
+* [invariants](./invariants/README.md): Define and exercise invariants in your code
```
Comment (patrick-ogrady) [inline README.md]: fix alphabetical order

File: codec/src/lib.rs [hunk 7]
```diff
@@ -215,10 +215,12 @@ commonware_macros::stability_scope!(BETA {
     pub use extensions::*;
 });
 
-#[cfg(feature = "arbitrary")]
-pub mod conformance;
+commonware_macros::stability_scope!(BETA {
```
Comment (patrick-ogrady) [inline codec/src/lib.rs]: Why does this need to be exported?

File: coding/src/lib.rs [hunk 8]
```diff
@@ -362,29 +373,41 @@ mod test {
...
-    proptest! {
-        #![proptest_config(ProptestConfig::with_cases(64))]
-
-        #[test]
-        fn proptest_roundtrip_reed_solomon(
-            (config, data, selected) in roundtrip_strategy()
-        ) {
-            roundtrip::<ReedSolomon<Sha256>>(&config, &data, &selected);
-        }
+    // Reed-Solomon requires extra_shards >= 1 (i.e., total > min).
```
Comment (patrick-ogrady) [inline coding/src/lib.rs]: I think we can now remove this comment (as we use NonZero for all)?

File: coding/src/lib.rs [hunk 9]
```diff
@@ -341,29 +352,41 @@ mod test {
...
+    #[test]
+    fn minifuzz_roundtrip_reed_solomon() {
+        minifuzz::Builder::default()
+            .with_seed(0)
+            .with_search_limit(64)
+            .test(|u| {
+                let (config, data, selected) = generate_case(u, 1)?;
+                roundtrip::<ReedSolomon<Sha256>>(&config, &data, &selected);
+                Ok(())
+            });
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply coding/src/lib.rs:383]: Probably worth addressing ^

File: AGENTS.md [hunk 10]
```diff
@@ -172,10 +172,11 @@ cargo llvm-cov --workspace --lcov --output-path lcov.info
 
 ## Testing Strategy
 
-- Unit tests: Core logic validation
-- Integration tests: Cross-primitive interaction
-- Fuzz tests: Input validation and edge cases
-- MIRI tests: Memory safety verification for unsafe code
+- Property: Core logic and invariant validation (use "minifuzz" from the invariants crate)
+- Unit: Testing edge cases or particular scenarios.
```
Comment (patrick-ogrady) [inline AGENTS.md:176]: remove `.` for consistency

File: invariants/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "commonware-invariants"
+edition.workspace = true
+publish = true
+version.workspace = true
+license.workspace = true
+description = "Define and exercise invariants in your code"
```
Comment (patrick-ogrady) [inline invariants/Cargo.toml]: nit: `Define and exercise invariants.`

File: invariants/README.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# commonware-invariants
+
+[![Crates.io](https://img.shields.io/crates/v/commonware-invariants.svg)](https://crates.io/crates/commonware-invariants)
+[![Docs.rs](https://docs.rs/commonware-invariants/badge.svg)](https://docs.rs/commonware-invariants)
+
+Define and exercise invariants in your code
```
Comment (patrick-ogrady) [inline invariants/README.md]: nit: `Define and exercise invariants.`

File: invariants/README.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+
+## Status
+
+`commonware-invariants` is **ALPHA** software and is not yet recommended for production use. Developers should expect breaking changes and occasional instability.
```
Comment (patrick-ogrady) [inline invariants/README.md:10]: Use the new status language from other crates

File: invariants/src/minifuzz.rs [hunk 14]
```diff
@@ -0,0 +1,564 @@
...
+//! is less extensive than a fuzz or property test.
+//!
+//! When using fuzzing in Rust, you have control over how random bytes are turned
+//! into inputs, so they effectively act more like property tests. This module provides
+//! a simple fuzzer, intended for unit tests. Unlike an actual fuzzer, this harness
+//! does not use coverage information, and is not going to run for as long, so it will
+//! naturally find fewer bugs. However, it can find many bugs pretty quickly,
+//! and can usually do a much better job than a unit test.
+//!
+//! In places where you'd normally write a unit test, you should instead consider
```
Comment (patrick-ogrady) [inline invariants/src/minifuzz.rs]: nit: imho we don't want to encourage this as a replacement for specific edge case testing (as it only has probabilistic coverage) but a way to generalize an edge case?

File: invariants/src/minifuzz.rs [hunk 15]
```diff
@@ -0,0 +1,564 @@
...
+//!         // same test body
+//!         Ok(())
+//!     });
+//! ```
+//!
+//! Use `with_search_limit` or `with_search_time` to control how long the fuzzer runs.
+//!
+//! # Why minifuzz?
+//!
+//! Minifuzz's goal is to both make it easier to write property tests, and to make
```
Comment (patrick-ogrady) [inline invariants/src/minifuzz.rs:32]: Should we also clarify that it doesn't exercise invalid cases (like a fuzz test normally would)?

Comment (patrick-ogrady) [reply invariants/src/minifuzz.rs:32]: In which case, is `fuzz` the wrong word?

File: invariants/src/minifuzz.rs
Comment (patrick-ogrady) [inline invariants/src/minifuzz.rs:1]: Like `conformance`, I think we could call this `invariants` or just use `lib.rs`? I'm not sure what else you plan to add here?

File: invariants/src/minifuzz.rs [hunk 16]
```diff
@@ -0,0 +1,564 @@
...
+    const fn new(seed: u64) -> Self {
+        Self {
+            seed: (seed >> 32) as u32,
+            thread: seed as u32,
+            size: 0,
+        }
+    }
+
+    fn try_from_hex(s: &str) -> Option<Self> {
+        let s = s.strip_prefix("0x").unwrap_or(s);
```
Comment (patrick-ogrady) [inline invariants/src/minifuzz.rs]: We already have `hex` handling in utils

File: invariants/src/minifuzz.rs [hunk 17]
```diff
@@ -0,0 +1,564 @@
...
+        let start = prefix_len + (self.rng.next_u32() as usize) % (self.buf.len() - prefix_len);
+        let len = ((self.rng.next_u32() as usize) % (self.buf.len() - start)).max(1);
+        let delta = self.rng.next_u32() as u8;
+        for b in &mut self.buf[start..start + len] {
+            *b = b.wrapping_add(delta);
+        }
+    }
+
+    fn pick_strategy(&mut self) -> u32 {
+        let mut weights = STRATEGY_WEIGHTS;
```
Comment (patrick-ogrady) [inline invariants/src/minifuzz.rs:254]: do we need to assign a var here?

File: invariants/Cargo.toml
Comment (cronokirby) [reply invariants/Cargo.toml:1]: It's always tempting to add something to `utils`, but it's important to resist that temptation when there's actually a coherent domain starting to form. Having a different crate also lets us easily depend on this feature only for testing, improving build times.

File: coding/src/lib.rs (hunk 8 above)
Comment (cronokirby) [reply coding/src/lib.rs]: This was missed when we updated that, I'd prefer this change in a separate PR.

File: invariants/Cargo.toml (hunk 11 above)
Comment (cronokirby) [reply invariants/Cargo.toml]: image

File: invariants/src/minifuzz.rs (hunk 14 above)
Comment (cronokirby) [reply invariants/src/minifuzz.rs]: Yeah, this language would be better as "you should _additionally_ consider".

File: invariants/src/minifuzz.rs (hunk 15 above)
Comment (cronokirby) [reply invariants/src/minifuzz.rs:32]: Yes, I prefer to think of our existing fuzz suites as being more like property tests, because we have control over input generation.

File: invariants/src/minifuzz.rs
Comment (cronokirby) [reply invariants/src/minifuzz.rs:1]: In the future we might have strategies for particular types, more more advanced testing strategies, like whitebox fuzzing, symbolic checking, etc. Having a place for such contributions makes sense. If we were to collapse minifuzz down into lib.rs, it would make more sense, imo, to just have that be the crate itself, i.e. call it "commonware-minifuzz"

File: math/src/lib.rs (hunk 2 above)
Comment (cronokirby) [reply math/src/lib.rs:24]: Yeah, this is a good point, I need to make the fuzz module inside of poly.rs ALPHA.

File: invariants/src/minifuzz.rs (hunk 17 above)
Comment (cronokirby) [reply invariants/src/minifuzz.rs:254]: Yes, in the function you use different weights depending on whether or not bytes are unused.

File: codec/src/lib.rs (hunk 7 above)
Comment (cronokirby) [reply codec/src/lib.rs]: It was exported before; the reason is that many crates use the utilities in this module to generate conformance tests exercising the codec implementation.

File: coding/src/lib.rs (hunk 8 above)
Comment (cronokirby) [reply coding/src/lib.rs]: Eh, might as well do this here.

File: invariants/Cargo.toml
Comment (patrick-ogrady) [reply invariants/Cargo.toml:1]: Ah, this is outdated. Sorry.

Comment (patrick-ogrady) [reply invariants/Cargo.toml:1]: Agree this is the right place.

File: invariants/src/minifuzz.rs [hunk 18]
```diff
@@ -0,0 +1,603 @@
...
+                            tries >= limit || deadline.is_some_and(|d| Instant::now() >= d);
+                        if past_min && past_limit {
+                            break 'search;
+                        }
+                    }
+                }
+            }
+            branch = branch.next();
+            sampler.switch(branch);
+        }
```
[the remark this answers is not in the document]
Comment (cronokirby) [reply invariants/src/minifuzz.rs:434]: Yes, this is intentional. The tradeoff is stalling rather than reporting a false positive.
